packages/eslint-plugin/tests/rules/no-invalid-this.test.ts

```diff
@@ -0,0 +1,465 @@
+    '(function() { console.log(this); z(x => console.log(x, this)); }).apply(obj);',
+
+    // Class Instance Methods.
+    'class A {foo() { console.log(this); z(x => console.log(x, this)); }};',
+
+    // Array methods.
+
+    'Array.from([], function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.every(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.filter(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.find(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.findIndex(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.forEach(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.map(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    'foo.some(function() { console.log(this); z(x => console.log(x, this)); }, obj);',
+
+    // @this tag.
+
+    '/** @this Obj */ function foo() { console.log(this); z(x => console.log(x, this)); }',
+
+    'foo(/* @this Obj */ function() { console.log(this); z(x => console.log(x, this)); });',
+
+    '/**\n * @returns {void}\n * @this Obj\n */\nfunction foo() { console.log(this); z(x => console.log(x, this)); }',
+
+    'Ctor = function() { console.log(this); z(x => console.log(x, this)); }',
+
+    'function foo(Ctor = function() { console.log(this); z(x => console.log(x, this)); }) {}',
+
+    '[obj.method = function() { console.log(this); z(x => console.log(x, this)); }] = a',
+
+    // Static
+
+    'class A {static foo() { console.log(this); z(x => console.log(x, this)); }};',
+  ],
+
+  invalid: [
+    {
+      code: `
+      interface SomeType { prop: string }
+      function foo() {
+        this.prop;
+      }`,
+      errors: [error],
+    },
+    // Global.
+    {
+      code: 'console.log(this); z(x => console.log(x, this));',
+
+      errors: [error, error],
+    },
+    {
+      code: 'console.log(this); z(x => console.log(x, this));',
+      parserOptions: {
+        ecmaFeatures: { globalReturn: true },
+      },
+      errors: [error, error],
+    },
+
+    // IIFE.
+    {
+      code:
+        '(function() { console.log(this); z(x => console.log(x, this)); })();',
+
+      errors: [error, error],
+    },
+
+    // Just functions.
+    {
+      code:
+        'function foo() { console.log(this); z(x => console.log(x, this)); }',
+
+      errors: [error, error],
+    },
+    {
+      code:
+        'function foo() { console.log(this); z(x => console.log(x, this)); }',
+
+      options: [{ capIsConstructor: false }], // test that the option doesn't reverse the logic and mistakenly allows lowercase functions
+      errors: [error, error],
+    },
+    {
+      code:
+        'function Foo() { console.log(this); z(x => console.log(x, this)); }',
+
+      options: [{ capIsConstructor: false }],
+      errors: [error, error],
```

Comment:
and this one for `error`

![image](https://user-images.githubusercontent.com/26347874/78166396-559e1400-746a-11ea-90aa-b28a706e7259.png)


both are of `string` type but string no overlap issue